libcaf_core/tests/legacy/mixin/requester.cpp

```diff
@@ -363,4 +363,23 @@ TEST_CASE("GH-1299 regression blocking") {
   }
 }
 
+TEST_CASE("GH-698 regression responding to a request drops the strong "
+          "reference count") {
+  auto weak_client_ref = [this] {
+    return actor_cast<weak_actor_ptr>(
+      sys.spawn([this](event_based_actor* self) {
+        self->request(adding_server, std::chrono::seconds{10}, 1, 2)
+          .then([this](int r) { *result = r; });
+      }));
+  }();
+  CHECK_EQ(weak_client_ref->strong_refs, 1u);
+  run_once();
+  CHECK_EQ(weak_client_ref->strong_refs, 2u);
+  expect((int, int), from(weak_client_ref).to(adding_server).with(1, 2));
+  CHECK_EQ(weak_client_ref->strong_refs, 2u);
+  expect((int), from(adding_server).to(weak_client_ref));
+  CHECK_EQ(weak_client_ref->strong_refs, 0u);
+  CHECK_EQ(*result, result_type{3});
```

Comment:
I don't get why we are tracking the reference count here if the issue was about timeouts not being cancelled. I would have expected something like this:

```
  run_once();
  expect((int, int), from(client).to(adding_server).with(1, 2));
  CHECK(sched.has_pending_timeout());
  expect((int), from(adding_server).to(client));
  CHECK(!sched.has_pending_timeout());
```

I would also expect a second test that covers the error case, i.e., if the client receives an error instead of an actual result.

Reply:
Ahh! Yeah, now it makes sense. Sorry for being a bit slow.

To make the test a bit more "digestible", I'd recommend explicitly triggering the `request` to start with a ref count of 1. Something like this:

```c++
  auto server = sys.spawn(...);
  auto client = actor_cast<strong_actor_ptr>(sys.spawn(...));
  run();
  CHECK_EQ(client->strong_refs, 1u);
  inject((actor), with(server).to(client));
  CHECK(sched.has_pending_timeout());
  expect((int, int), from(client).to(server));
  expect((int), from(server).to(client));
  CHECK(!sched.has_pending_timeout());
  // The scheduler may no longer hold a reference to the client.
  CHECK_EQ(client->strong_refs, 1u);
```